core/src/main/java/com/linecorp/armeria/client/Http2ResponseDecoder.java

```diff
@@ -257,22 +260,17 @@ public int onDataRead(
         final long writtenBytes = res.writtenBytes();
         if (maxContentLength > 0 && writtenBytes > maxContentLength - dataLength) {
             final long transferred = LongMath.saturatedAdd(writtenBytes, dataLength);
-            res.close(ContentTooLargeException.builder()
-                                              .maxContentLength(maxContentLength)
-                                              .contentLength(res.headers())
-                                              .transferred(transferred)
-                                              .build());
+            res.close(contentTooLargeException(res, transferred));
             throw connectionError(
                     INTERNAL_ERROR,
                     "content too large: transferred(%d + %d) > limit(%d) (stream: %d)",
                     writtenBytes, dataLength, maxContentLength, streamId);
         }
 
-        try {
-            res.write(HttpData.wrap(data.retain()).withEndOfStream(endOfStream));
-        } catch (Throwable t) {
-            res.close(t);
-            throw connectionError(INTERNAL_ERROR, t, "failed to consume a DATA frame");
+        if (!res.tryWriteData(HttpData.wrap(data.retain()).withEndOfStream(endOfStream))) {
+            final ClosedSessionException cause = ClosedSessionException.get();
+            res.close(cause);
```

Comment:
Not related to this PR; If `tryWriteData()` returns false, it means `res` was closed already. Do we need to call `res.close(cause)`?

Reply:
That's correct. We don't have to call it. Thanks!